feat: add `password_sha256_hash ` for `user` resource to support OpenTofu

MattiasMTS opened the pull request:
Fulfills #95

MattiasMTS commented:
> Hello @MattiasMTS. Is there a specific reason for submitting this PR when #95 addresses the same issue?

Hi @nunoadrego ! There is no specific reason at all - I just want this in now 😄 
This PR adds e.g. `stringvalidator.ExactlyOneOf` (which I see you've already reviewed the other PR with)

So I am all for merging either - would be very nice to get this feature in ASAP 🙇

nunoadrego commented on pkg/resource/user/model.go:
```diff
@@ -5,9 +5,10 @@ import (
+	ClusterName                   types.String `tfsdk:"cluster_name"`
+	ID                            types.String `tfsdk:"id"`
+	Name                          types.String `tfsdk:"name"`
+	PasswordHash                  types.String `tfsdk:"password_hash"`
```
```suggestion
	PasswordHash                  types.String `tfsdk:"password_sha256_hash"`
```
For consistency, what do you think?

nunoadrego commented on pkg/resource/user/user.go:
```diff
@@ -61,23 +61,42 @@ func (r *Resource) Schema(_ context.Context, _ resource.SchemaRequest, resp *res
 					int32planmodifier.RequiresReplace(),
 				},
 			},
+			"password_hash": schema.StringAttribute{
```
```suggestion
			"password_sha256_hash": schema.StringAttribute{
```

nunoadrego reviewed (commented):
I added some comments for you to consider. We can try to make some progress in this PR, since the https://github.com/ClickHouse/terraform-provider-clickhousedbops/pull/95 is waiting for feedback from the contributor.

MattiasMTS commented on docs/resources/user.md:
```diff
@@ -4,49 +4,72 @@ page_title: "clickhousedbops_user Resource - clickhousedbops"
 This field must be left null when using a ClickHouse Cloud cluster.
 When using a self hosted ClickHouse instance, this field should only be set when there is more than one replica and you are not using 'replicated' storage for user_directory.
 - `host_ips` (Set of String) IP addresses from which the user is allowed to connect. If not specified, user can connect from any host.
+- `password_sha256_hash` (String, Sensitive, Deprecated) SHA256 hash of the password to be set for the user. Use this field for OpenTofu compatibility. Changes to this field will trigger a resource replacement.
```
good idea

MattiasMTS commented on pkg/resource/user/user.go:
```diff
@@ -61,23 +61,42 @@ func (r *Resource) Schema(_ context.Context, _ resource.SchemaRequest, resp *res
 			},
 			"password_sha256_hash_wo_version": schema.Int32Attribute{
-				Required:    true,
+				Optional:    true,
```
I updated it so you've to use `password_sha256_hash_wo_version` in conjunction with `password_sha256_hash_wo` (see 2nd test case in `TestUser_validation_acceptance`) to ensure backward compatibility with the current version (where both fields are required)

MattiasMTS commented:
@nunoadrego  thanks again for your thorough review and suggestions!! ✨ I believe I've addressed all of your comments now. Would be awesome if you could take a final look 🙇

nunoadrego commented on pkg/resource/user/user.md:
```diff
@@ -1,7 +1,17 @@
-- When importing an existing user, the `clickhousedbops_user` resource will be lacking the `password_sha256_hash_wo_version` and thus the subsequent apply will need to recreate the database User in order to set a password.
+- Changing the password will cause the database user to be deleted and recreated.
+- Changing `password_sha256_hash_wo` alone does not trigger an update. You must also bump `password_sha256_hash_wo_version`.
+- When importing an existing user configured with legacy fields, the resource will be lacking the `password_sha256_hash_wo_version` and the subsequent apply will recreate the user to set a password.
```
```suggestion
- When importing an existing user, the `clickhousedbops_user` resource will be lacking the password or the `password_sha256_hash_wo_version`, and thus the subsequent apply will need to recreate the database User in order to set a password.
```

nunoadrego commented on pkg/resource/user/user_test.go:
```diff
@@ -147,7 +150,78 @@ func TestUser_acceptance(t *testing.T) {
+					Config: fmt.Sprintf(`
+					resource "clickhousedbops_user" "test" {
+						name                    = "testuser"
+						password_sha256_hash    = "%s"
```
Do you think this is needed? Either the description should be changed or this should be removed.

MattiasMTS commented:
> Looks good to me, @MattiasMTS! Thank you for your contribution. We will now run the E2E tests and release an alpha version on Monday 🙇 

Thank you! 🙏 great review ✨🙌
